packages/react-router-dom/index.tsx

```diff
@@ -976,6 +977,95 @@ export function useFormAction(
   return createPath(path);
 }
 
+let blockerKey = "blocker-singleton";
+
+export function useBlocker(
+  shouldBlock: boolean | (() => boolean) | BlockerFunction
+) {
+  let { router } = useDataRouterContext(DataRouterHook.UseFetcher);
+
+  let blockerFunction = React.useCallback<BlockerFunction>(() => {
+    return typeof shouldBlock === "function"
+      ? shouldBlock() === true
+      : shouldBlock === true;
+  }, [shouldBlock]);
+
+  let blocker = router.getBlocker(blockerKey, blockerFunction);
+
+  // Cleanup on unmount
+  React.useEffect(() => () => router.deleteBlocker(blockerKey), [router]);
+
+  return blocker;
+}
+
+export function usePrompt(
+  message: string | null | false,
+  opts?: { beforeUnload: boolean }
+) {
+  let { beforeUnload } = opts ? opts : { beforeUnload: false };
+  let blockerFunction = React.useCallback<BlockerFunction>(() => {
+    if (!message) {
+      return false;
+    }
+
+    // Here be dragons! This is not bulletproof (at least at the moment).  If
+    // you click the back button again while the global window.confirm prompt
+    // is open, it returns `false` (telling us to "block") but then _also_
+    // processes the back button click!
+
+    // So, consider:
+    //  - you have a stack of [/a, /b, /c] and you usePrompt() on /c
+    //  - user clicks back button trying to POP /c -> /b
+    //  - prompt shows up (URL shows /b but UI shows /c)
+    //  - user clicks back button _again_ (POP /b -> /a)
+    //    - this seems to queue up internally
+    //  - we get a `false` back from window.confirm() (block!)
+    //  - so we undo the _original_ POP /c -> /b and call history.go(1) to
+    //    instead POP forward /b -> /c and we also tell our router to ignore
+    //    the next history update
+    //  - and then it seems that the queued history trumps our history revert,
+    //    and so we receive the popstate event for the POP /b -> /a and then
+    //    we ignore it thinking it was our revert of POP /b -> /c
+    //
+    //  I think the solution here is to track more thn a boolean
+    //  ignoreNextHistoryUpdate and instead track the key we're reverting from
+    //  and the delta and compare that to any incoming popstate events.
+    return !window.confirm(message);
+  }, [message]);
+
+  let blocker = useBlocker(blockerFunction);
+
+  let prevState = React.useRef(blocker.state);
+  React.useEffect(() => {
+    if (blocker.state === "blocked") {
```

Comment:
I think this logic isn't quite correct with the way things are currently written. AFAICT this will make it so that usePrompt is only able to prevent navigation once, where really you want to prevent until either the back-button is clicked twice or the user confirms the prompt.